File: Programmers/[42892] 길 찾기 게임/KingDonggyu/42892.js
const Node = (function () {
  function Node(x, number) {
    this.x = x;
    this.number = number;
    this.left = null;
    this.right = null;
  }

  Node.prototype.insert = function (x, number) {
    this.x > x 
			? this.addLeft(x, number) 
			: this.addRight(x, number);
  };

  Node.prototype.addLeft = function (x, number) {
    this.left 
			? this.left.insert(x, number) 
			: (this.left = new Node(x, number));
  };

  Node.prototype.addRight = function (x, number) {
    this.right
      ? this.right.insert(x, number)
      : (this.right = new Node(x, number));
  };

  return Node;
})();

function solution(nodeinfo) {
  const nodes = nodeinfo
    .map(([x, y], i) => ({ x, y, number: i + 1 }))
    .sort((a, b) => b.y - a.y);

  const root = new Node(nodes[0].x, nodes[0].number);

  for (let i = 1; i < nodes.length; i++) {
    root.insert(nodes[i].x, nodes[i].number);
  }

  const answer = [[], []];

  const preorder = (node) => {
    if (node) {
      answer[0].push(node.number);
      preorder(node.left);
      preorder(node.right);
    }
  };

  const postorder = (node) => {
    if (node) {
      postorder(node.left);
      postorder(node.right);
      answer[1].push(node.number);
    }
  };

  preorder(root);
  postorder(root);

  return answer;
}

File: Programmers/[42892] 길 찾기 게임/ONE/42892.java
import java.util.ArrayList;
import java.util.List;

class Node {
    int num;
    int x;
    int y;
    Node left;
    Node right;

    public Node(int num, int x, int y, Node left, Node right) {
        this.num = num;
        this.x = x;
        this.y = y;
        this.left = left;
        this.right = right;
    }

    public void insert(Node parent, Node child) {
        if (parent.x > child.x) {
            if (parent.left == null)
                parent.left = child;
            else
                insert(parent.left, child);
        }
        else {
            if (parent.right == null)
                parent.right = child;
            else
                insert(parent.right, child);
        }
    }
}

class Solution {
    private int[][] answer;
    private int index;
    public int[][] solution(int[][] nodeinfo) {
        answer = new int[2][nodeinfo.length];

        List<Node> tree = new ArrayList<>();
        for(int i = 0; i < nodeinfo.length; i++)
            tree.add(new Node(i + 1, nodeinfo[i][0], nodeinfo[i][1], null, null));

        tree.sort((o1, o2) -> {
            if (o1.y == o2.y)
                return o1.x - o2.x;
            return o2.y - o1.y;
        });

        Node root = tree.get(0);

        for(int i = 1; i < tree.size(); i++)
            root.insert(root, tree.get(i));

        postorder(root);
        index = 0;
        preorder(root);

        return answer;
    }

    private void preorder(Node node) {
        if (node != null) {
            answer[0][index++] = node.num;
            preorder(node.left);
            preorder(node.right);
        }
    }

    private void postorder(Node node) {
        if (node != null) {
            postorder(node.left);
            postorder(node.right);
            answer[1][index++] = node.num;
        }
    }
}